File: src/type/v2/timeline.ts
/**
 * Timeline型定義（v2）
 *
 * 物語の時系列を管理するための型定義
 * ハイブリッド詳細管理（インライン + ファイル参照）に対応
 */

/**
 * イベントのカテゴリ
 *
 * - `plot_point`: プロットの転換点
 * - `character_event`: キャラクター関連イベント
 * - `world_event`: 世界観に関するイベント
 * - `backstory`: 過去の出来事
 * - `foreshadow`: 伏線となるイベント
 * - `climax`: クライマックス
 * - `resolution`: 解決・結末
 */
export type EventCategory =
  | "plot_point"
  | "character_event"
  | "world_event"
  | "backstory"
  | "foreshadow"
  | "climax"
  | "resolution";

/**
 * イベントの重要度
 *
 * - `major`: 物語の主要イベント
 * - `minor`: 補助的なイベント
 * - `background`: 背景イベント
 */
export type EventImportance = "major" | "minor" | "background";

/**
 * タイムラインのスコープ
 *
 * - `story`: 物語全体のタイムライン
 * - `world`: 世界観のタイムライン
 * - `character`: キャラクター個別のタイムライン
 * - `arc`: ストーリーアーク単位のタイムライン
 */
export type TimelineScope = "story" | "world" | "character" | "arc";

/**
 * 時間点（TimePoint）
 *
 * イベントの発生時点を表現する
 */
export type TimePoint = {
  /** 順序（必須）：タイムライン内での相対的な順序 */
  order: number;
  /** ラベル（オプショナル）：表示用の時間表記 */
  label?: string;
  /** 日付（オプショナル）：物語内での日付表記 */
  date?: string;
  /** チャプター（オプショナル）：関連するチャプターID */
  chapter?: string;
};

/**
 * イベントの詳細情報（オプショナル）
 */
export type TimelineEventDetails = {
  /** 詳細説明 */
  description?: string | { file: string };
  /** イベントの影響 */
  impact?: string | { file: string };
  /** メモ */
  notes?: string | { file: string };
};

/**
 * イベントによるキャラクターフェーズ変化
 * このイベントをきっかけにキャラクターがフェーズ遷移する情報
 */
export type PhaseChangeInfo = {
  /** キャラクターID */
  characterId: string;
  /** 遷移先フェーズID */
  toPhaseId: string;
  /** 遷移元フェーズID（オプション） */
  fromPhaseId?: string;
  /** 説明（オプション） */
  description?: string;
};

/**
 * LSP用の検出ヒント（イベント用）
 */
export type TimelineEventDetectionHints = {
  /** よく使われるパターン */
  commonPatterns: string[];
  /** 除外すべきパターン */
  excludePatterns: string[];
  /** 検出の信頼度（0.0～1.0） */
  confidence: number;
};

/**
 * タイムラインイベント型
 *
 * 物語内で発生する個別のイベントを表現
 */
export type TimelineEvent = {
  // ========================================
  // 必須メタデータ
  // ========================================

  /** 一意なID（プログラム的な識別子） */
  id: string;

  /** イベントタイトル */
  title: string;

  /** イベントカテゴリ */
  category: EventCategory;

  /** 発生時点 */
  time: TimePoint;

  /** 短い概要（必須） */
  summary: string;

  /** 関連するキャラクターのIDリスト */
  characters: string[];

  /** 関連する設定（場所など）のIDリスト */
  settings: string[];

  /** 関連するチャプターのIDリスト */
  chapters: string[];

  // ========================================
  // オプショナル情報
  // ========================================

  /** このイベントの原因となったイベントのIDリスト */
  causedBy?: string[];

  /** このイベントが引き起こすイベントのIDリスト */
  causes?: string[];

  /** イベントの重要度 */
  importance?: EventImportance;

  /** イベント終了時点（期間があるイベントの場合） */
  endTime?: TimePoint;

  /** 表示名のバリエーション */
  displayNames?: string[];

  /** 詳細情報 */
  details?: TimelineEventDetails;

  /** LSP用の検出ヒント */
  detectionHints?: TimelineEventDetectionHints;

  /**
   * このイベントによるキャラクターフェーズ変化
   * イベントをフェーズ遷移のトリガーとしてリンクする
   */
  phaseChanges?: PhaseChangeInfo[];
};

/**
 * タイムラインの詳細情報（オプショナル）
 */
export type TimelineDetails = {
  /** 背景説明 */
  background?: string | { file: string };
  /** メモ */
  notes?: string | { file: string };
};

/**
 * LSP用の検出ヒント（タイムライン用）
 */
export type TimelineDetectionHints = {
  /** よく使われるパターン */
  commonPatterns: string[];
  /** 除外すべきパターン */
  excludePatterns: string[];
  /** 検出の信頼度（0.0～1.0） */
  confidence: number;
};

/**
 * タイムラインの表示オプション
 */
export type TimelineDisplayOptions = {
  /** 関連を表示するか */
  showRelations?: boolean;
  /** カラースキーム */
  colorScheme?: string;
  /** 折りたたみ状態 */
  collapsed?: boolean;
};

/**
 * タイムライン型
 *
 * 物語の時系列全体を管理するコンテナ
 */
export type Timeline = {
  // ========================================
  // 必須メタデータ
  // ========================================

  /** 一意なID（プログラム的な識別子） */
  id: string;

  /** タイムライン名 */
  name: string;

  /** タイムラインのスコープ */
  scope: TimelineScope;

  /** 短い概要（必須） */
  summary: string;

  /** イベントリスト */
  events: TimelineEvent[];

  // ========================================
  // オプショナル情報
  // ========================================

  /** 親タイムラインのID */
  parentTimeline?: string;

  /** 子タイムラインのIDリスト */
  childTimelines?: string[];

  /** 関連するキャラクターのID（character scopeの場合） */
  relatedCharacter?: string;

  /** 表示名のバリエーション */
  displayNames?: string[];

  /** 表示オプション */
  displayOptions?: TimelineDisplayOptions;

  /** 詳細情報 */
  details?: TimelineDetails;
};

File: internal/domain/timeline.go
package domain

// Timeline / TimelineEvent / TimePoint mirror src/type/v2/timeline.ts.
// Bundled in one file because TimelineEvent only ever exists inside Timeline,
// and the LSP cache walks them together — splitting per type would force
// circular imports for shared enum constants.

// TimelineScope enumerates the four timeline scopes from timeline.ts.
//
// Why: typed string instead of int — preserves wire-format parity with the
// existing TS YAML/JSON snapshots without extra (Un)Marshal hooks.
type TimelineScope string

const (
	TimelineScopeStory     TimelineScope = "story"
	TimelineScopeWorld     TimelineScope = "world"
	TimelineScopeCharacter TimelineScope = "character"
	TimelineScopeArc       TimelineScope = "arc"
)

// EventCategory enumerates timeline event categories. Values mirror TS
// `EventCategory` in src/type/v2/timeline.ts.
type EventCategory string

const (
	EventCategoryPlotPoint      EventCategory = "plot_point"
	EventCategoryCharacterEvent EventCategory = "character_event"
	EventCategoryWorldEvent     EventCategory = "world_event"
	EventCategoryBackstory      EventCategory = "backstory"
	EventCategoryForeshadow     EventCategory = "foreshadow"
	EventCategoryClimax         EventCategory = "climax"
	EventCategoryResolution     EventCategory = "resolution"
)

// EventImportance ranks how central an event is to the narrative.
type EventImportance string

const (
	EventImportanceMajor      EventImportance = "major"
	EventImportanceMinor      EventImportance = "minor"
	EventImportanceBackground EventImportance = "background"
)

// TimePoint anchors an event in narrative time. Order is required; the rest
// stay optional via pointer so a zero TimePoint is valid.
type TimePoint struct {
	// Required: relative ordering within the parent timeline.
	Order int `json:"order"`

	// Optional display / linkage metadata.
	Label   *string `json:"label,omitempty"`
	Date    *string `json:"date,omitempty"`
	Chapter *string `json:"chapter,omitempty"`
}

// PhaseChangeInfo records a character phase transition triggered by an event.
type PhaseChangeInfo struct {
	CharacterID string  `json:"characterId"`
	ToPhaseID   string  `json:"toPhaseId"`
	FromPhaseID *string `json:"fromPhaseId,omitempty"`
	Description *string `json:"description,omitempty"`
}

// TimelineEventDetails carries optional long-form fields that may live inline
// or be split out into Markdown via the shared StringOrFileRef helper.
type TimelineEventDetails struct {
	Description *StringOrFileRef `json:"description,omitempty"`
	Impact      *StringOrFileRef `json:"impact,omitempty"`
	Notes       *StringOrFileRef `json:"notes,omitempty"`
}

// TimelineEventDetectionHints feeds the LSP detector with patterns specific
// to this event.
type TimelineEventDetectionHints struct {
	CommonPatterns  []string `json:"commonPatterns"`
	ExcludePatterns []string `json:"excludePatterns"`
	Confidence      float64  `json:"confidence"`
}

// TimelineEvent is one beat on a timeline. Cross-entity references
// (Characters, Settings, Chapters, CausedBy, Causes) stay as string IDs to
// keep the domain package free of inter-entity pointer cycles.
type TimelineEvent struct {
	// ----- required metadata -----
	ID         string        `json:"id"`
	Title      string        `json:"title"`
	Category   EventCategory `json:"category"`
	Time       TimePoint     `json:"time"`
	Summary    string        `json:"summary"`
	Characters []string      `json:"characters"`
	Settings   []string      `json:"settings"`
	Chapters   []string      `json:"chapters"`

	// ----- optional fields -----
	CausedBy       []string                     `json:"causedBy,omitempty"`
	Causes         []string                     `json:"causes,omitempty"`
	Importance     *EventImportance             `json:"importance,omitempty"`
	EndTime        *TimePoint                   `json:"endTime,omitempty"`
	DisplayNames   []string                     `json:"displayNames,omitempty"`
	Details        *TimelineEventDetails        `json:"details,omitempty"`
	DetectionHints *TimelineEventDetectionHints `json:"detectionHints,omitempty"`
	PhaseChanges   []PhaseChangeInfo            `json:"phaseChanges,omitempty"`
}

// TimelineDetails carries optional long-form Timeline fields.
type TimelineDetails struct {
	Background *StringOrFileRef `json:"background,omitempty"`
	Notes      *StringOrFileRef `json:"notes,omitempty"`
}

// TimelineDetectionHints feeds the LSP detector with patterns for the
// timeline label itself.
type TimelineDetectionHints struct {
	CommonPatterns  []string `json:"commonPatterns"`
	ExcludePatterns []string `json:"excludePatterns"`
	Confidence      float64  `json:"confidence"`
}

// TimelineDisplayOptions controls UI rendering hints from timeline.ts.
type TimelineDisplayOptions struct {
	ShowRelations *bool   `json:"showRelations,omitempty"`
	ColorScheme   *string `json:"colorScheme,omitempty"`
	Collapsed     *bool   `json:"collapsed,omitempty"`
}

// Timeline is the container for an ordered list of TimelineEvents within a
// given Scope. Parent / child links and RelatedCharacter remain string IDs
// so the LSP cache can resolve them lazily.
type Timeline struct {
	// ----- required metadata -----
	ID      string          `json:"id"`
	Name    string          `json:"name"`
	Scope   TimelineScope   `json:"scope"`
	Summary string          `json:"summary"`
	Events  []TimelineEvent `json:"events"`

	// ----- optional fields -----
	ParentTimeline   *string                 `json:"parentTimeline,omitempty"`
	ChildTimelines   []string                `json:"childTimelines,omitempty"`
	RelatedCharacter *string                 `json:"relatedCharacter,omitempty"`
	DisplayNames     []string                `json:"displayNames,omitempty"`
	DisplayOptions   *TimelineDisplayOptions `json:"displayOptions,omitempty"`
	Details          *TimelineDetails        `json:"details,omitempty"`
	DetectionHints   *TimelineDetectionHints `json:"detectionHints,omitempty"`
}
